### DISCUSSION TITLE
Default Optimization settings for Modes

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/greenwood#545] [General]
As Greenwood has the concepts of [_mode_](https://www.greenwoodjs.io/docs/configuration/#mode) and [optimization](https://www.greenwoodjs.io/docs/configuration/#optimization):
- _mode_: the type of app your building, eg. SSG vs SPA and more about project architecture
- _optimization_: how you want to deliver your assets, either statically, or preloaded, etc

It would be good to explore and provide / document defaults for each mode.
|Mode|JavaScript|CSS|Router|GraphQL
|-----|-----|-----|-----|-----|
|SSG|`defer`|`inline`|N / A|`defer`|
|MPA|`defer`|`inline`|`defer`|`defer`|
|SPA|`preload`|`preload`|`defer`|`async`|

**Comment by thescientist13** [comment]
Noticing these warning with the current settings, fwiw
<img width="1673" alt="preload-warnings" src="https://user-images.githubusercontent.com/895923/121790593-e9ab5580-cbae-11eb-8e86-c219dfa39e31.png">

And looks like Next is moving to `defer`?
https://twitter.com/shubhie/status/1391891309338660864

**Reply by thescientist13** [reply]
We should make sure to take any notes from [this video](https://youtu.be/Lh9q3h2khlc), in particular some takeaways 
- should still keep JavaScript loading at the end, or at least use defer
- base64 encoding can be more bulky / increase payloads
- keep assets (images, SVGs) out of JS as much as possible, e.g. don't bundle 
- can use link preload for images!

**Comment by thescientist13** [comment]
Good post by Guy Bedford covering code / splitting and module preloading .
https://guybedford.com/es-module-preloading-integrity

**Comment by thescientist13** [comment]
As part of working on #546 , I wasn't able to get a `data-gwd-opt="none"` override option to work, since for [JavaScript (Rollup + Terser)](https://github.com/ProjectEvergreen/greenwood/blob/v0.13.0/packages/cli/src/plugins/resource/plugin-standard-javascript.js#L42) and for [CSS (PostCSS)](https://github.com/ProjectEvergreen/greenwood/blob/master/packages/cli/src/plugins/resource/plugin-standard-css.js#L52), optimizations happen at a "plugin" level for those respective tools, which means it's applied to all files equally.  So somehow we would need to figure out a way to filter or exclude certain files from this somehow, but likely that would have to be done at the plugin level, and until we build the app, we don't know which files to exclude, and of course it will keep changing as development happens.

Only immediate thought I had was that maybe we could use sourcemaps to reverse engineer the source?  But we don't necessarily want the source, we want the bundled output, but just nothing more than that, for just that particular file... 🤔 

----

edit: actually, for CSS, it's done per file, so we might have a better chance supporting it that way, but how would be able to hint from the browser that is has a particular `data` attribute??

**Comment by thescientist13** [comment]
An interesting note from [this thread](https://github.com/whatwg/html/issues/2791#issuecomment-892798322)
> _JS modules do not block the parser. `<script type="module">` implies `defer`._

**Comment by thescientist13** [comment]
This could be a useful (optional) diagnostic tool to provide, maybe?
https://csswizardry.com/ct/

**Reply by aholtzman** [reply]
Indeed. It may be good to see where we can optimize how static files are generated.
